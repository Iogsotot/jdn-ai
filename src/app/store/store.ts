import { configureStore } from '@reduxjs/toolkit';
import undoable from 'redux-undo';

import mainSlice from '../main.slice';
import filterSlice from '../../features/filter/filter.slice';
import locatorsSlice from '../../features/locators/locators.slice';
import customLocatorSlice from '../../features/locators/customLocator.slice';
import onboardingSlice from '../../features/onboarding/store/onboarding.slice';
import pageObjectSlice from '../../features/pageObjects/pageObject.slice';

import { scriptNotifier } from '../../pageServices/scriptNotifier';
import { updateMessageHandler } from '../../pageServices/scriptMessageHandler';

import { updateSocketMessageHandler } from '../../services/webSocketMessageHandler';

// import { logger } from './middlewares/logger';
import { changePageMiddleware } from './middlewares/changePage.middleware';
import { shouldRunGeneration } from '../../features/locators/reducers/shouldRunGeneration.middleware';
import { cancellableActions } from '../../common/components/notification/middlewares/cancellableActions';

import { onLocatorsCreated } from '../../features/locators/reducers/identifyElements.thunk';

import { quitThrottlerMiddleware } from '../../common/utils/throttler';

import { createLogger } from 'redux-logger';

const logger = createLogger();

const rootReducer = {
  main: mainSlice,
  filters: filterSlice,
  locators: undoable(locatorsSlice, { undoType: 'LOCATOR_UNDO', jumpType: 'LOCATOR_JUMP' }),
  pageObject: undoable(pageObjectSlice, { undoType: 'PAGEOBJECT_UNDO' }),
  onboarding: onboardingSlice,
  customLocator: customLocatorSlice,
};

export const store = configureStore({
  reducer: rootReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: [
          'locators/setElementGroupGeneration',
          'locators/elementSetActive',
          'locators/elementGroupSetActive',
          'locators/setActiveSingle',
          'locators/updateLocatorGroup',
          'main/setScriptMessage',
        ],
      },
    }).concat([
      logger,
      scriptNotifier,
      cancellableActions,
      changePageMiddleware,
      shouldRunGeneration,
      quitThrottlerMiddleware,
      onLocatorsCreated,
    ]),
});

store.subscribe(() => updateMessageHandler(store.dispatch, store.getState()));
store.subscribe(() => updateSocketMessageHandler(store.dispatch, store.getState()));

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
