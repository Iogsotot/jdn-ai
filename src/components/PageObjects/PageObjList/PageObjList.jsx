import React, { useEffect, useState } from "react";
import { useSelector } from "react-redux";
import { Collapse } from "antd";
import { size } from "lodash";
import Icon from "@ant-design/icons";

import { PageObjListHeader } from "./PageObjListHeader";
import { PageObjMenu } from "./PageObjMenu";

import CaretDownSvg from "../../../assets/caret-down.svg";
import PageSvg from "../../../assets/page.svg";
import { selectConfirmedLocators, selectPageObjects } from "../../../store/selectors/pageObjectSelectors";
import { Locator } from "../../Locators/Locator";
import { GenerationButtons } from "./GenerationButtons";
import { PageObjectPlaceholder } from "../PageObjectPlaceholder";
import { PageObjCopyButton } from "./PageObjCopyButton";

export const PageObjList = (props) => {
  const state = useSelector((state) => state);
  const currentPageObject = useSelector((state) => state.pageObject.currentPageObject);
  const pageObjects = useSelector(selectPageObjects);
  const [activePanel, setActivePanel] = useState([currentPageObject]);

  useEffect(() => {
    setActivePanel([currentPageObject]);
  }, [currentPageObject]);

  const renderLocators = (elements) => {
    if (size(elements)) {
      return elements.map((element) => <Locator key={element.element_id} {...{ element }} noScrolling={true} />);
    } else {
      return "No locators selected";
    }
  };

  const renderPageObjSettings = (pageObjId, url) => {
    return (
      <div className="jdn__pageObject__settings">
        <div className="jdn__pageObject__settings-url">{url}</div>
        <GenerationButtons pageObj={pageObjId} />
      </div>
    );
  };

  const renderContent = (pageObjId, url, elements) => {
    if (size(elements)) {
      return renderLocators(elements);
    } else {
      return renderPageObjSettings(pageObjId, url);
    }
  };

  return (
    <div className="jdn__locatorsList">
      <PageObjListHeader {...props} />
      <div className="jdn__locatorsList-content jdn__pageObj-content">
        {size(pageObjects) ? (
          <Collapse
            expandIcon={({ isActive }) => (
              <Icon component={CaretDownSvg} rotate={isActive ? 180 : 270} fill="#808080" />
            )}
            expandIconPosition="start"
            activeKey={activePanel}
            onChange={setActivePanel}
          >
            {pageObjects.map(({ id, name, url, locators, library }) => {
              const elements = selectConfirmedLocators(state, id);
              const isPageObjectNotEmpty = !!size(elements);
              return (
                <Collapse.Panel
                  key={id}
                  header={
                    <React.Fragment>
                      <Icon component={PageSvg} className="jdn__locatorsList-status" />
                      <span className="jdn__pageObject-content-text">{name}</span>
                    </React.Fragment>
                  }
                  extra={
                    <>
                      {isPageObjectNotEmpty && <PageObjCopyButton {...{ elements }} />}
                      <PageObjMenu {...{ id, name, locators, elements, library }} />
                    </>
                  }
                >
                  {renderContent(id, url, elements)}
                </Collapse.Panel>
              );
            })}
          </Collapse>
        ) : (
          <PageObjectPlaceholder />
        )}
      </div>
    </div>
  );
};
