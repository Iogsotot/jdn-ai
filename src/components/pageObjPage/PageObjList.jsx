import React from "react";
import { useSelector } from "react-redux";
import Icon from "@ant-design/icons";
import { Collapse } from "antd";

import { PageObjListHeader } from "./PageObjListHeader";

import CaretDownSvg from "../../assets/caret-down.svg";
import PageSvg from "../../assets/page.svg";
import { selectPageObjects } from "../../store/selectors/pageObjectSelectors";

export const PageObjList = () => {
  const pageObjects = useSelector(selectPageObjects);

  return (
    <div className="jdn__locatorsList">
      <PageObjListHeader />
      <div className="jdn__locatorsList-content">
        <Collapse
          className="jdn__collapse"
          expandIcon={({ isActive }) => <Icon component={CaretDownSvg} rotate={isActive ? 180 : 0} />}
        >
          {/* {renderPanel()} */}
          {pageObjects.map(({ id, name }) => (
            <Collapse.Panel
              key={id}
              header={
                <React.Fragment>
                  <Icon component={PageSvg} className="jdn__locatorsList-status" />
                  {name}
                </React.Fragment>
              }
              className="jdn__collapse-panel"
            >
              {"Locators list 1"}
            </Collapse.Panel>
          ))}
        </Collapse>
      </div>
    </div>
  );
};
