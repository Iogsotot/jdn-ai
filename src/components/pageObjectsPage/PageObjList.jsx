import React, { useState, useEffect } from "react";
import { useSelector } from "react-redux";
import Icon from "@ant-design/icons";
import { Collapse } from "antd";
import { size } from "lodash";

import { PageObjListHeader } from "./PageObjListHeader";

import CaretDownSvg from "../../assets/caret-down.svg";
import PageSvg from "../../assets/page.svg";
import { selectConfirmedLocators, selectPageObjects } from "../../store/selectors/pageObjectSelectors";
import { Locator } from "../locatorsPage/Locator";
import { GenerationButtons } from "../GenerationButtons";

export const PageObjList = () => {
  const state = useSelector((state) => state);
  const currentPageObject = useSelector((state) => state.pageObject.currentPageObject);
  const pageObjects = useSelector(selectPageObjects);
  const xpathConfig = useSelector((state) => state.main.xpathConfig);
  const [activePanel, setActivePanel] = useState([]);

  useEffect(() => {
    setActivePanel([...activePanel, currentPageObject]);
  }, [currentPageObject]);

  const renderLocators = (pageObjId) => {
    const elements = selectConfirmedLocators(state, pageObjId);
    if (size(elements)) {
      return elements.map((element) => <Locator key={element.element_id} {...{ element, xpathConfig }} />);
    } else {
      return "No locators selected";
    }
  };

  const renderPageObjSettings = (pageObjId, url) => {
    return (
      <div className="jdn__pageObject__settings">
        <div className="jdn__pageObject__settings-url">{url}</div>
        <GenerationButtons pageObj={pageObjId} />
      </div>
    );
  };

  return (
    <div className="jdn__locatorsList">
      <PageObjListHeader />
      <div className="jdn__locatorsList-content jdn__pageObj-content">
        <Collapse
          className="jdn__collapse"
          expandIcon={({ isActive }) => <Icon component={CaretDownSvg} rotate={isActive ? 180 : 0} />}
          activeKey={activePanel}
          onChange={setActivePanel}
        >
          {pageObjects.map(({ id, name, url, locators }) => (
            <Collapse.Panel
              key={id}
              header={
                <React.Fragment>
                  <Icon component={PageSvg} className="jdn__locatorsList-status" />
                  {name}
                </React.Fragment>
              }
              className="jdn__collapse-panel"
            >
              {size(locators) ? renderLocators(id) : renderPageObjSettings(id, url)}
            </Collapse.Panel>
          ))}
        </Collapse>
      </div>
    </div>
  );
};
