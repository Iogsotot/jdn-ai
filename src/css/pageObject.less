@import "./variables.less";

.jdn {
  &__pageObject {
    &__settings {
      background-color: #ecf7ff;
      padding: 12px 52px 16px;

      &-url {
        color: #808080;
        font-size: 12px;
      }
    }

    & .jdn__xpath_container--shift {
      padding-left: 48px;
    }

    & .ant-collapse > .ant-collapse-item > .ant-collapse-header .ant-collapse-arrow {
      margin-right: 8px;
    }

    &_placeholder {
      display: flex;
      flex-direction: column;

      .anticon svg {
        height: 50px;
        width: 50px;
        margin-top: 40px;
        margin-bottom: 24px;
      }

      .ant-typography {
        text-align: center;
        margin-bottom: 16px;
        font-size: 12px;
      }
    }
  }
}
