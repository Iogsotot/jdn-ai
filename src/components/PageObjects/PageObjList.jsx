import React, { useState, useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import Icon from "@ant-design/icons";
import { Collapse, Dropdown, Menu, Typography } from "antd";
import { size } from "lodash";

import { PageObjListHeader } from "./PageObjListHeader";

import CaretDownSvg from "../../assets/caret-down.svg";
import PageSvg from "../../assets/page.svg";
import EllipsisSvg from "../../assets/ellipsis.svg";
import TrashBinSvg from "../../assets/trash-bin.svg";
import DownloadSvg from "../../assets/download.svg";
import PencilSvg from "../../assets/pencil.svg";
import { selectConfirmedLocators, selectPageObjects } from "../../store/selectors/pageObjectSelectors";
import { Locator } from "../Locators/Locator";
import { GenerationButtons } from "./GenerationButtons";
import { PageObjectPlaceholder } from "./PageObjectPlaceholder";
import { removePageObject } from "../../store/slices/pageObjectSlice";
import { removeLocators } from "../../store/slices/locatorsSlice";
import { generatePageObject } from "../../services/pageObject";
import { pageType } from "../../utils/constants";
import { changePage } from "../../store/slices/mainSlice";

export const PageObjList = () => {
  const state = useSelector((state) => state);
  const dispatch = useDispatch();
  const currentPageObject = useSelector((state) => state.pageObject.currentPageObject);
  const pageObjects = useSelector(selectPageObjects);
  const [activePanel, setActivePanel] = useState([]);

  useEffect(() => {
    setActivePanel([currentPageObject]);
  }, [currentPageObject]);

  const renderLocators = (elements) => {
    if (size(elements)) {
      return elements.map((element) => <Locator key={element.element_id} {...{ element }} noScrolling={true} />);
    } else {
      return "No locators selected";
    }
  };

  const renderPageObjSettings = (pageObjId, url) => {
    return (
      <div className="jdn__pageObject__settings">
        <div className="jdn__pageObject__settings-url">{url}</div>
        <GenerationButtons pageObj={pageObjId} />
      </div>
    );
  };

  const renderContent = (pageObjId, url, elements) => {
    if (size(elements)) {
      return renderLocators(elements);
    } else {
      return renderPageObjSettings(pageObjId, url);
    }
  };

  const handleRename = (id, name) => {
    chrome.storage.sync.set({ OPEN_EDIT_NAME: { isOpen: true, value: {id, name} } });
  };

  const renderMenu = (id, locatorIds, locatorObjects, name) => {
    const handleRemove = () => {
      dispatch(removePageObject(id));
      dispatch(removeLocators(locatorIds));
    };

    const handleDownload = () => {
      generatePageObject(locatorObjects, name);
    };

    const handleEdit = () => {
      dispatch(changePage({ page: pageType.locatorsList, pageObj: currentPageObject, alreadyGenerated: true }));
    };

    return (
      <Menu>
        <Menu.Item key="3" icon={<PencilSvg />} onClick={() => handleRename(id, name)}>
          Rename
        </Menu.Item>
        {size(locatorIds) ? (
          <React.Fragment>
            <Menu.Item key="4" icon={<PencilSvg />} onClick={handleEdit}>
              Edit list
            </Menu.Item>
            <Menu.Item key="5" icon={<DownloadSvg />} onClick={handleDownload}>
              Download
            </Menu.Item>
          </React.Fragment>
        ) : null}
        <Menu.Item key="6" icon={<TrashBinSvg />} onClick={handleRemove}>
          <Typography.Text type="danger">Delete</Typography.Text>
        </Menu.Item>
      </Menu>
    );
  };

  return (
    <div className="jdn__locatorsList">
      <PageObjListHeader />
      <div className="jdn__locatorsList-content jdn__pageObj-content">
        {size(pageObjects) ? (
          <Collapse
            expandIcon={({ isActive }) => (
              <Icon component={CaretDownSvg} rotate={isActive ? 180 : 270} fill="#808080" />
            )}
            expandIconPosition="left"
            activeKey={activePanel}
            onChange={setActivePanel}
          >
            {pageObjects.map(({ id, name, url, locators }) => {
              const elements = selectConfirmedLocators(state, id);
              return (
                <Collapse.Panel
                  key={id}
                  header={
                    <React.Fragment>
                      <Icon component={PageSvg} className="jdn__locatorsList-status" />
                      {name}
                    </React.Fragment>
                  }
                  extra={
                    <a onClick={(e) => e.stopPropagation()} data-testid="dropdown-button">
                      <Dropdown trigger="click" overlay={renderMenu(id, locators, elements, name)}>
                        <Icon component={EllipsisSvg} />
                      </Dropdown>
                    </a>
                  }
                >
                  {renderContent(id, url, elements)}
                </Collapse.Panel>
              );
            })}
          </Collapse>
        ) : (
          <PageObjectPlaceholder />
        )}
      </div>
    </div>
  );
};
