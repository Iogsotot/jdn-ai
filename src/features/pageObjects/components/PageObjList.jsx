import Icon from "@ant-design/icons";
import { Collapse, Tooltip, Typography } from "antd";
import { size } from "lodash";
import React, { useEffect, useState } from "react";
import { useSelector } from "react-redux";

import { CaretDown } from "phosphor-react";
import PageSvg from "../assets/page.svg";
import { Footnote } from "../../../common/components/footnote/Footnote";
import { selectConfirmedLocators, selectPageObjects } from "../pageObject.selectors";
import { GenerationButton } from "./GenerationButton";
import { PageObjectPlaceholder } from "./PageObjectPlaceholder";
import { PageObjCopyButton } from "./PageObjCopyButton";
import { Locator } from "../../locators/Locator";
import { PageObjMenu } from "./PageObjMenu";
import { PageObjListHeader } from "./PageObjListHeader";
import { Notifications } from "../../../common/components/notification/Notifications";

export const PageObjList = (props) => {
  const state = useSelector((state) => state);
  const currentPageObject = useSelector((state) => state.pageObject.present.currentPageObject);
  const pageObjects = useSelector(selectPageObjects);
  const [activePanel, setActivePanel] = useState([currentPageObject]);

  const isExpanded = !!size(activePanel);

  useEffect(() => {
    setActivePanel([currentPageObject]);
  }, [currentPageObject]);

  const renderLocators = (elements) => {
    if (size(elements)) {
      return elements.map((element) => <Locator key={element.element_id} {...{ element }} noScrolling={true} />);
    } else {
      return "No locators selected";
    }
  };

  const renderPageObjGeneration = (pageObjId, url, library) => {
    return (
      <div className="jdn__pageObject__settings">
        <Footnote className="jdn__pageObject__settings-url">{url}</Footnote>
        <GenerationButton pageObj={pageObjId} {...{ library }} />
      </div>
    );
  };

  const renderContent = (pageObjId, url, elements, library) => {
    if (size(elements)) {
      return renderLocators(elements);
    } else {
      return renderPageObjGeneration(pageObjId, url, library);
    }
  };

  const toggleExpand = () => {
    if (size(activePanel)) {
      setActivePanel([]);
    } else {
      const keys = pageObjects.map((po) => po.id);
      setActivePanel(keys);
    }
  };

  return (
    <div className="jdn__locatorsList">
      <PageObjListHeader {...{ ...props, toggleExpand, isExpanded, setActivePanel }} />
      <div className="jdn__locatorsList-content jdn__pageObject-content">
        {size(pageObjects) ? (
          <React.Fragment>
            <Collapse
              expandIcon={({ isActive }) => (
                <CaretDown
                  style={{
                    transform: isActive ? "rotate(180deg)" : "rotate(0deg)",
                  }}
                  size={14}
                  color="#878A9C"
                />
              )}
              expandIconPosition="start"
              activeKey={activePanel}
              onChange={setActivePanel}
            >
              {pageObjects.map(({ id, name, url, locators, library }) => {
                const elements = selectConfirmedLocators(state, id);
                const isPageObjectNotEmpty = !!size(elements);
                return (
                  <Collapse.Panel
                    key={id}
                    header={
                      <Tooltip
                        title={url}
                        placement="bottomLeft"
                        getPopupContainer={(triggerNode) => triggerNode}
                        align={{ offset: [-28, 0] }}
                      >
                        <Icon component={PageSvg} className="jdn__locatorsList-status" />
                        <Typography.Text className="jdn__pageObject-content-text">{name}</Typography.Text>
                      </Tooltip>
                    }
                    extra={
                      <>
                        {isPageObjectNotEmpty && <PageObjCopyButton {...{ elements }} />}
                        <PageObjMenu {...{ id, name, locators, elements, library }} />
                      </>
                    }
                  >
                    {renderContent(id, url, elements, library)}
                  </Collapse.Panel>
                );
              })}
            </Collapse>
          </React.Fragment>
        ) : (
          <PageObjectPlaceholder />
        )}
        <Notifications />
      </div>
    </div>
  );
};
