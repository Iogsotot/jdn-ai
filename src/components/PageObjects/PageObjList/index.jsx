import { PageObjList as _PageObjList } from "./PageObjList";

export const PageObjList = _PageObjList;
