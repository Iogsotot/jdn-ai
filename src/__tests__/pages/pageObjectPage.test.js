import React from "react";
import "@testing-library/jest-dom";

import { cleanup, render, screen, waitForElementToBeRemoved } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { Provider } from "react-redux";
import { jest } from "@jest/globals";

import { store } from "../../store/store";
import { PageObjList } from "../../components/PageObjects/PageObjList";
import * as pageObject from "../../services/pageObject";

describe("application launch", () => {
  const TestComponent = () => (
    <Provider {...{ store }}>
      <PageObjList />
    </Provider>
  );

  beforeAll(() => {
    jest.mock("@ant-design/icons");
    jest.spyOn(pageObject, "getPageAttributes").mockImplementation(() => [
      {
        result: {
          title: "HomePage",
          url: "https://jdi-testing.github.io/jdi-light/contacts.html",
        },
      },
    ]);
  });

  afterEach(() => cleanup());

  test("renders properly for empty PO list", () => {
    const { container } = render(<TestComponent />);
    expect(container).toMatchSnapshot();
  });

  test("new PO is added by clicking button", async () => {
    render(<TestComponent />);
    await userEvent.click(screen.getByRole("button"));
    await userEvent.click(screen.getByRole("button"));
    await waitForElementToBeRemoved(() => screen.getByText("There are no created page objects."));
    const items = await screen.findAllByText(/HomePage/);
    expect(items).toHaveLength(2);
  });
});
