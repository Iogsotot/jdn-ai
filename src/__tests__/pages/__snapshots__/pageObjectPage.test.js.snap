// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`application launch renders properly for empty PO list 1`] = `
<div>
  <div
    class="jdn__locatorsList"
  >
    <div
      class="jdn__locatorsList-header"
    >
      <span
        class="jdn__locatorsList-header-title"
      >
        Page Objects
      </span>
      <div
        class="jdn__locatorsList-header-buttons"
      >
        <button
          class="ant-btn jdn__buttons"
          type="button"
        >
          <span
            class="anticon"
            role="img"
          />
          <span>
            New page object
          </span>
        </button>
      </div>
    </div>
    <div
      class="jdn__locatorsList-content jdn__pageObj-content"
    >
      <div
        class="jdn__pageObject_placeholder"
      >
        <span
          class="anticon"
          role="img"
        />
        <article
          class="ant-typography"
        >
          There are no created page objects.
        </article>
        <article
          class="ant-typography"
        >
          Open the needed page and click the "New page object" button to get started.
        </article>
      </div>
    </div>
  </div>
</div>
`;
